### PR TITLE
Forms

### PR DESCRIPTION
**Pull request by Angelica-zb**
1. Task: https://github.com/rolling-scopes-school/tasks/tree/master/react/modules/module02
2. Screenshot:
image
image

3. Deploy: https://spontaneous-hamster-76b21e.netlify.app/forms
4. Done 26.03.2023 / deadline 27.03.2023
5. Score: 15 /15

### CODE REVIEW

File: .eslintrc.json [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    "extends": [
+        "eslint:recommended",
+        "plugin:react/recommended",
+        "plugin:@typescript-eslint/recommended",
+        "prettier",
+        "plugin:react-hooks/recommended",
+        "plugin:prettier/recommended"
+    ],
+    "root": true,
+    "overrides": [
```
Comment (Ksarelto) [inline .eslintrc.json]: remove it

File: .eslintrc.json [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        "prettier/prettier": ["error", { "endOfLine": "auto" }],
+        "@typescript-eslint/interface-name-prefix": "off",
+        "@typescript-eslint/ban-ts-comment": "error",
+        "@typescript-eslint/no-non-null-assertion": "off",
+        "@typescript-eslint/explicit-module-boundary-types": "off",
+        "@typescript-eslint/no-empty-function": "off",
+        "@typescript-eslint/no-explicit-any": "error",
+        "@typescript-eslint/no-var-reqiures": "off",
+        "react/jsx-uses-react": "off",
+        "react/react-in-jsx-scope": "off"
```
Comment (Ksarelto) [inline .eslintrc.json]: sort rules: react rules => typescript rules => other rules

File: .prettierrc [hunk 3]
```diff
@@ -0,0 +1,9 @@
+{
+    "endOfLine": "auto",
+    "semi": true,
+    "singleQuote": true,
+    "tabWidth": 2,
+    "trailingComma": "es5",
+    "printWidth": 100,
+    "arrowParens": "always"
+}
```
Comment (Ksarelto) [inline .prettierrc:9]: add empty lines in the end of all files

File: .gitignore [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+# Editor directories and files
+.vscode/*
+!.vscode/extensions.json
+.idea
+.DS_Store
+*.suo
+*.ntvs*
+*.njsproj
+*.sln
+*.sw?
```
Comment (Ksarelto) [inline .gitignore:25]: add 'coverage' here

File: src/App.scss [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    }
+}
+
+.card {
+    padding: 2em;
+}
+
+.read-the-docs {
+    color: #888;
+}
```
Comment (Ksarelto) [inline src/App.scss:42]: add empty line

File: src/App.tsx [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+import { Routes } from 'react-router-dom';
+import { Route } from 'react-router';
+import Header from './components/header/header';
+import AboutUs from './pages/aboutUs/aboutUs-page';
+import Main from './pages/main/main-page';
+import FormsPage from './pages/forms/forms-page';
+import Error from './pages/error/error-page';
+
+const App = () => {
+  return (
```
Comment (Ksarelto) [inline src/App.tsx:11]: u can omit return

File: src/components/allCards/allCards.tsx [hunk 7]
```diff
@@ -0,0 +1,21 @@
+import { ICard } from 'interfaces/ICardInterfaces';
+import React from 'react';
+import Card from '../card/card';
+import classes from './allCards.module.scss';
+
+interface IAllCards {
+  cards: ICard[];
+}
+class AllCards extends React.Component<IAllCards, object> {
```
Comment (Ksarelto) [inline src/components/allCards/allCards.tsx:10]: empty line before class

File: src/components/card/card.module.scss [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+        background-color: $back-input;
+    }
+    &:hover .card__image {
+        transition: all 0.7s;
+        transform: scale(1.1);
+    }
+    .cardImg {
+        height: 60%;
+    }
+    .descriptionCard {
```
Comment (Ksarelto) [inline src/components/card/card.module.scss:34]: spaces before every style block

File: src/components/card/card.test.tsx [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+describe('Card', () => {
+  const card = {
+    id: 1,
+    name: 'Яблоки',
+    cost: 12,
+    description: 'Яблоки спелые, очень красивые',
+    image: 'https://m.dom-eda.com/uploads/images/catalog/item/86df51de21/c25c94fe96_500.jpg',
+  };
+
+  it('render correct img', () => {
```
Comment (Ksarelto) [inline src/components/card/card.test.tsx]: add 'should' in the begining

File: vite.config.ts [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+    coverage: {
+      enabled: true,
+      provider: 'c8',
+      reporter: ['text'],
+      all: true,
+      include: ['src'],
+      exclude: ['src/types'],
+    },
+  },
+});
```
Comment (Ksarelto) [inline vite.config.ts:21]: empty line in the end

File: src/pages/forms/forms-page.tsx [hunk 11]
```diff
@@ -0,0 +1,113 @@
+import { IErrors, IForms } from '../../interfaces/ICardInterfaces';
```
Comment (Ksarelto) [inline src/pages/forms/forms-page.tsx:1]: files with components and its folders should start from Capital letters

File: src/pages/forms/forms-page.tsx [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+    };
+  }
+
+  sendValid = (objData: IForms) => {
+    const { errors, isValid } = this.valedation(objData);
+
+    if (isValid) {
+      alert('Вы в списке');
+      const copyState = this.state.dataValid;
+      copyState.push(objData);
```
Comment (Ksarelto) [inline src/pages/forms/forms-page.tsx]: mutation of state => very very dengerous

File: src/pages/forms/forms-page.tsx [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+        });
+      }, 1000);
+    } else {
+      this.setState({
+        errors: errors,
+      });
+    }
+  };
+
+  valedation = (objData: IForms) => {
```
Comment (Ksarelto) [inline src/pages/forms/forms-page.tsx]: validation

File: src/pages/forms/forms-page.tsx [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+      male: true,
+      agreement: true,
+    };
+
+    if (objData.name.length > 3) {
+      errors.name = false;
+    } else {
+      isValid = false;
+    }
+    if (objData.foto !== '') {
```
Comment (Ksarelto) [inline src/pages/forms/forms-page.tsx]: empty string is converted to false so u can check as :
```
if(objData.foto) {}
```

File: src/pages/forms/forms-page.tsx [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+  valedation = (objData: IForms) => {
+    let isValid = true;
+    const errors: IErrors = {
+      name: true,
+      foto: true,
+      birth: true,
+      fruit: true,
+      male: true,
+      agreement: true,
+    };
```
Comment (Ksarelto) [inline src/pages/forms/forms-page.tsx:62]: u can create something like:
```
const { name, foto, birth, fruit, male, agreement } = objData

const errors: IErrors = {
      name: name.length > 3,
      foto: !!foto,
      birth: new Date(birth).getTime() < new Date().getTime(),
      fruit: !!fruit,
      male: !!male,
      agreement: !!agreement,
 };

const isValid = Object.values(errors).every(Boolean)
```

File: src/pages/error/error.test.tsx [hunk 16]
```diff
@@ -0,0 +1,10 @@
+import React from 'react';
```
Comment (Ksarelto) [inline src/pages/error/error.test.tsx]: no need in this import

File: src/interfaces/ICardInterfaces.tsx [hunk 17]
```diff
@@ -0,0 +1,25 @@
+export interface ICard {
```
Comment (Ksarelto) [inline src/interfaces/ICardInterfaces.tsx:1]: folder is already named interfaces, so don't add interfaces into the name of the file

File: src/components/userCards/userCards.tsx [hunk 18]
```diff
@@ -0,0 +1,21 @@
+import React from 'react';
+import classes from '../userCards/userCards.module.scss';
+import { IForms } from '../../interfaces/ICardInterfaces';
+import UserCard from '../userCard/userCard';
+// import Forms from '../../components/formsBlank/forms';
```
Comment (Ksarelto) [inline src/components/userCards/userCards.tsx]: remove comment, and styles int he bottom of the list of imports

File: src/components/userCards/userCards.tsx [hunk 19]
```diff
@@ -0,0 +1,21 @@
...
+
+class UserCards extends React.Component<{ allCards: Array<IForms> }> {
+  render() {
+    if (this.props.allCards.length > 0) {
```
Comment (Ksarelto) [inline src/components/userCards/userCards.tsx]: if (this.props.allCards.length) {}

File: src/components/userCard/userCard.test.tsx [hunk 20]
```diff
@@ -0,0 +1,25 @@
...
+    agreement: true,
+  };
+
+  it('renders correctly', () => {
+    const { getByAltText, getByText } = render(<UserCard formaValid={mockFormData} />);
+    expect(getByAltText('foto')).toBeInTheDocument();
+    expect(getByText('Имя:John Doe')).toBeInTheDocument();
+    expect(getByText('Пол: male')).toBeInTheDocument();
+    expect(getByText('Дата рождения: 1990-01-01')).toBeInTheDocument();
+    expect(getByText('Любимый фрукт: apple')).toBeInTheDocument();
```
Comment (Ksarelto) [inline src/components/userCard/userCard.test.tsx:22]: to many expects

File: src/components/userCard/userCard.module.scss [hunk 21]
```diff
@@ -0,0 +1,13 @@
...
+    width: 20%;
+    display: flex;
+    flex-direction: column;
+    background-color: rgba(59, 88, 152, 0.3);
+    border-radius: 10px;
+    margin: 10px;
+    padding: 10px;
+    img {
+        height: 200px;
+        // height: 50px;
```
Comment (Ksarelto) [inline src/components/userCard/userCard.module.scss]: remove comment

File: src/components/search/search.module.scss [hunk 22]
```diff
@@ -0,0 +1,23 @@
+.search {
+    margin: 15px 0;
+    input {
```
Comment (Ksarelto) [inline src/components/search/search.module.scss:3]: spaces between styles and try to use classes for styles

File: src/components/formsBlank/forms.tsx [hunk 23]
```diff
@@ -0,0 +1,132 @@
...
+      birth: '',
+      fruit: '',
+      male: '',
+      agreement: false,
+    };
+    event.preventDefault();
+    if (this.nameRef.current?.value) {
+      objData.name = this.nameRef.current.value;
+    }
+    if (this.fotoRef.current?.value) {
```
Comment (Ksarelto) [inline src/components/formsBlank/forms.tsx:38]: empty line before if blocks

File: src/components/formsBlank/forms.tsx [hunk 24]
```diff
@@ -0,0 +1,132 @@
...
+    if (this.maleRef.current?.checked) {
+      objData.male = this.maleRef.current.value;
+    }
+    if (this.femaleRef.current?.checked) {
+      objData.male = this.femaleRef.current.value;
+    }
+    if (this.agreementRef.current?.value) {
+      objData.agreement = this.agreementRef.current.checked;
+    }
+    console.log(objData);
```
Comment (Ksarelto) [inline src/components/formsBlank/forms.tsx]: do you need this console?

File: src/components/formsBlank/forms.tsx [hunk 25]
```diff
@@ -0,0 +1,132 @@
...
+
+  submitForm = (event: React.ChangeEvent<HTMLFormElement>) => {
+    const objData = {
+      name: '',
+      foto: '',
+      birth: '',
+      fruit: '',
+      male: '',
+      agreement: false,
+    };
```
Comment (Ksarelto) [inline src/components/formsBlank/forms.tsx:31]: u can do as:
```
const objData = {
      name: this.nameRef.current?.value || '',
      foto: this.fotoRef.current?.value || '',
      birth: this.birthRef.current?.value || '',
      fruit: this.fruitRef.current?.value || '',
      male: this.maleRef.current?.value || this.femaleRef.current?.value || '',
      agreement: this.agreementRef.current?.value || false,
    };
```

File: src/App.tsx (hunk 6 above)
Comment (Angelica-zb) [reply src/App.tsx:11]: i can't. it broke page

**Pull request merged**